### PR TITLE
cmd/core/ip

### LINKED ISSUE #3012: ip
# IP Command

## Usage

```bash
ip [ OPTIONS ] OBJECT { COMMAND | help }

ip [ -force ] -batch filename

where  OBJECT := { address | addrlabel | amt | fou | help | ila | ioam | l2tp | link | macsec | maddress | monitor | mptcp | mroute | mrule | neighbor | neighbour | netconf | netns | nexthop | ntable | ntbl | route | rule | sr | tap | tcpmetrics | token | tunnel | tuntap | vrf | xfrm }


OPTIONS := { -V[ersion] | -s[tatistics] | -d[etails] | -r[esolve] | -h[uman-readable] | -iec | -j[son] | -p[retty] | -f[amily] { inet | inet6 | mpls | bridge | link } | -4 | -6 | -M | -B | -0 | -l[oops] { maximum-addr-flush-attempts } | -br[ief] | -o[neline] | -t[imestamp] | -ts[hort] | -b[atch] [filename] | -rc[vbuf] [size] | -n[etns] name | -N[umeric] | -a[ll] | -c[olor] | -echo}
```

## Progress

 Commands and options are checked if atleast one subcommand exist.

Options

- [ ] -V[ersion]
- [ ] -s[tatistics] / -stats
- [ ] -d[etails]
- [ ] -r[esolve]
- [ ] -h[uman-readable]
- [ ] -iec
- [ ] -j[son]
- [ ] -p[retty]
- [ ] -f[amily]
- [ ] -4
- [ ] -6
- [ ] -M
- [ ] -B
- [ ] -0
- [ ] -l[oops]
- [ ] -br[ief]
- [ ] -o[neline]
- [ ] -t[imestamp]
- [ ] -ts[hort]
- [ ] -b[atch] [filename]
- [ ] -rc[vbuf] [size]
- [ ] -n[etns] name
- [ ] -N[umeric]
- [ ] -a[ll]
- [ ] -c[olor]
- [ ] -echo

Command

- [x] address
  - [x] add
  - [ ] change
  - [ ] replace
  - [x] del
  - [ ] show
  - [ ] save
  - [ ] flush
  - [ ] showdump
  - [ ] restore

- [ ] addrlabel
  - [ ] add
  - [ ] del
  - [ ] list
  - [ ] flush
  - [ ] help

- [ ] fou
  - [ ] add
  - [ ] del
  - [ ] show

- [ ] ila?
  - [ ] add
  - [ ] del
  - [ ] list
  - [ ] help

- [ ] ioam
  - [ ] namespace
    - [ ] add
    - [ ] delete
    - [ ] show
    - [ ] set
  - [ ] schema
    - [ ] add
    - [ ] delete
    - [ ] show
  - [ ] help

- [ ] l2tp
  - [ ] add
    - [ ] tunnel
    - [ ] session
  - [ ] del
    - [ ] tunnel
    - [ ] session
  - [ ] show
    - [ ] tunnel
    - [ ] session

- [x] link
  - [x] add
  - [ ] delete
  - [x] set
  - [x] show
  - [ ] xstats
  - [ ] afstats
  - [ ] property
    - [ ] add
    - [ ] del
  - [ ] help

- [ ] macsec
  - [ ] add
  - [ ] set
  - [ ] del
  - [ ] show
  - [ ] offload
  - [ ] help

- [ ] maddr(ess)
  - [ ] add
  - [ ] del
  - [ ] show
  - [ ] help

- [ ] monitor
  - [ ] all
  - [ ] address
  - [ ] link
  - [ ] mroute
  - [ ] neigh
  - [ ] netconf
  - [ ] nexthop
  - [ ] nsid
  - [ ] prefix
  - [ ] route
  - [ ] rule
  - [ ] help
  
- [ ] mptcp
  - [ ] endpoint
    - [ ] add
    - [ ] delete
    - [ ] show
    - [ ] flush
  - [ ] limit
    - [ ] set
    - [ ] show
  - [ ] monitor
  - [ ] help

- [ ] mroute
  - [ ] show
  - [ ] help

- [x] neigh(bour/bor)
  - [ ] add
  - [ ] del
  - [ ] change
  - [ ] replace
  - [x] show
  - [ ] flush
  - [ ] get

- [ ] netconf
  - [ ] show
  - [ ] help

- [ ] netns
  - [ ] list
  - [ ] add
  - [ ] attach
  - [ ] set
  - [ ] delete
  - [ ] identify
  - [ ] pids
  - [ ] exec
  - [ ] monitor
  - [ ] list-id
  - [ ] help

- [ ] nexthop
  - [ ] list
  - [ ] flush
  - [ ] add
  - [ ] replace
  - [ ] get
  - [ ] del
  - [ ] bucket
    - [ ] list
    - [ ] get
  - [ ] help

- [ ] ntable
  - [ ] change
  - [ ] show
  - [ ] help

- [x] route
  - [ ] list
  - [ ] flush
  - [ ] save
  - [ ] restore
  - [ ] showdump
  - [ ] get
  - [x] add
  - [x] del
  - [ ] change
  - [ ] append
  - [ ] replace
  - [ ] help

- [ ] rule / mrule
  - [ ] list
  - [ ] add
  - [ ] del
  - [ ] flush
  - [ ] save
  - [ ] restore
  - [ ] help

- [ ] sr  
  - [ ] hmac
    - [ ] set
    - [ ] show
  - [ ] tunsrc
    - [ ] set
    - [ ] show
  - [ ] help

- [ ] tcp_metrics
  - [ ] show
  - [ ] flush
  - [ ] delete
  - [ ] help

- [ ] token
  - [ ] list
  - [ ] set
  - [ ] del
  - [ ] get
  - [ ] help

- [ ] tunnel
  - [ ] add
  - [ ] change
  - [ ] del
  - [ ] show
  - [ ] prl
  - [ ] 6rd
  - [ ] help

- [ ] tuntap / tap
  - [ ] add
  - [ ] del
  - [ ] show
  - [ ] list
  - [ ] lst
  - [ ] help

- [ ] vrf
  - [ ] show
  - [ ] exec
  - [ ] identify
  - [ ] pids
  - [ ] help

- [ ] xfrm
  - [ ] state
    - [ ] add
    - [ ] update
    - [ ] allocspi
    - [ ] delete
    - [ ] get
    - [ ] deleteall
    - [ ] list
    - [ ] flush
    - [ ] count
  - [ ] policy
    - [ ] add
    - [ ] update
    - [ ] delete
    - [ ] get
    - [ ] deleteall
    - [ ] list
    - [ ] flush
    - [ ] count
    - [ ] set
    - [ ] setdefault
    - [ ] getdefault
  - [ ] monitor
    - [ ] all
    - [ ] aquire
    - [ ] expire
    - [ ] SA
    - [ ] aevent
    - [ ] policy
    - [ ] report
  - [ ] help

### PR DESCRIPTION
**Pull request by RiSKeD**
resolves #3012 
# Todos
- [x] prioritize feature set
- [ ] implement missing feature set 

Implementation Progress
# IP Command

## Usage

```bash
ip [ OPTIONS ] OBJECT { COMMAND | help }

ip [ -force ] -batch filename

where  OBJECT := { address | addrlabel | amt | fou | help | ila | ioam | l2tp | link | macsec | maddress | monitor | mptcp | mroute | mrule | neighbor | neighbour | netconf | netns | nexthop | ntable | ntbl | route | rule | sr | tap | tcpmetrics | token | tunnel | tuntap | vrf | xfrm }


OPTIONS := { -V[ersion] | -s[tatistics] | -d[etails] | -r[esolve] | -h[uman-readable] | -iec | -j[son] | -p[retty] | -f[amily] { inet | inet6 | mpls | bridge | link } | -4 | -6 | -M | -B | -0 | -l[oops] { maximum-addr-flush-attempts } | -br[ief] | -o[neline] | -t[imestamp] | -ts[hort] | -b[atch] [filename] | -rc[vbuf] [size] | -n[etns] name | -N[umeric] | -a[ll] | -c[olor] | -echo}
```

## Implementation Status

 Commands and options are checked if atleast one subcommand exist.

Command

- [x] address
  - [x] add
  - [ ] change
  - [x] replace
  - [x] del
  - [x] show
  - [x] flush

- [x] link
  - [x] add
  - [x] delete
  - [x] set
  - [x] show
  - [ ] xstats
  - [ ] afstats

- [x] neigh(bour/bor)
  - [x] add
  - [x] del
  - [ ] change
  - [x] replace
  - [x] show
  - [x] flush
  - [x] get

- [x] route
  - [x] list
  - [x] flush
  - [x] get
  - [x] add
  - [x] del
  - [ ] change
  - [x] append
  - [x] replace

- [x] monitor
  - [x] all
  - [x] address
  - [x] link
  - [ ] mroute
  - [x] neigh
  - [ ] netconf
  - [ ] nexthop
  - [ ] nsid
  - [ ] prefix
  - [x] route
  - [ ] rule

- [x] tunnel
  - [x] add
  - [ ] change
  - [x] del
  - [x] show

- [x] tuntap / tap
  - [x] add
  - [x] del
  - [x] show
  - [x] list
  - [x] lst

- [ ] token
  - [ ] list
  - [ ] set
  - [ ] del
  - [ ] get

- [x] vrf
  - [x] show
  - [ ] exec
  - [ ] identify
  - [ ] pids

- [ ] tcp_metrics
  - [x] show
  - [ ] flush
  - [ ] delete

- [ ] nexthop
  - [ ] list
  - [ ] flush
  - [ ] add
  - [ ] replace
  - [ ] get
  - [ ] del
  - [ ] bucket
    - [ ] list
    - [ ] get

- [x] xfrm
  - [x] state
    - [x] add
    - [x] update
    - [x] allocspi
    - [x] delete
    - [x] get
    - [x] deleteall
    - [x] list
    - [x] flush
    - [x] count
  - [x] policy
    - [x] add
    - [x] update
    - [x] delete
    - [x] get
    - [x] deleteall
    - [x] list
    - [x] flush
    - [x] count
    - [ ] set
  - [x] monitor
    - [x] all
    - [x] aquire
    - [x] expire
    - [x] SA
    - [x] aevent
    - [x] policy
    - [x] report

### Options

- [x] -V[ersion]
- [x] -s[tatistics] / -stats
- [x] -d[etails]
- [x] -r[esolve]
- [x] -h[uman-readable]
- [x] -iec
- [x] -j[son]
- [x] -p[retty]
- [x] -f[amily]
- [x] -4
- [x] -6
- [x] -M
- [x] -B
- [x] -0
- [x] -l[oops]
- [x] -br[ief]
- [x] -o[neline]
- [x] -t[imestamp]
- [x] -ts[hort]
- [x] -b[atch] [filename]
- [x] -rc[vbuf] [size]
- [x] -n[etns] name
- [x] -N[umeric]
- [x] -a[ll]

### Testing

- [x] unit tests for conversions
- [ ] integration tests

### CODE REVIEW

**Review by rminnich — CHANGES_REQUESTED** [review]
I would prefer that you not make this change to parsing. 

The parsing has a very important goal: on error, the program should say exactly what it parsed successfully; what went wrong and why; and what was expected. 

While the current code could do this better, the current code is a HUGE improvement on the standard IP command, which is "dump an incorrect bnf on any error and ask the user to guess again". Further, the BNF that the ip command dumps is *wrong* -- the approach of the u-root IP command is nicer, because as the code changes, the error handling stays in sync. Further, you don't have to constantly see if there are enough args in argv -- if you panic due to running off the end of the array, it's ok, because the code is using panic in the parsing.

It may seem odd to use panic at all, given the general Go principle: "don't panic"
But rules are not absolute. I got the idea for using panic, as done here, from talks Rob Pike gave. I talked to Russ in 2018 about this usage of panic, and his comment is: "I still think it [panic] makes sense in a parser ..." with the caveat that it should not cross package boundaries. That rule is followed here (since it's all in main :-) and in the forth parser. Panic can make code a lot simpler.

Again, let's stick to extending the command, but I really don't want to see this parser rewrite.

**Comment by RiSKeD** [comment]
The panics motivated me to rewrite it, but I understand that the old parsing is better from a user perspective.

**Comment by 10000TB** [comment]
@RiSKeD should we make this ready for review ?

**Review by 10000TB — COMMENTED** [review]
left comments

**Comment by RiSKeD** [comment]
> @RiSKeD should we make this ready for review ?

Not yet, I am iterating over all the subcommands which can be integrated with https://github.com/vishvananda/netlink/tree/main at the moment. After that, we can start the review process and look into what else should be prioritized in implementation.

File: cmds/core/ip/address_linux.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	c := findPrefix(cmd, whatIWant)
+	switch c {
+	case "show":
+		return addressShow(w)
+	case "add", "change", "replace", "del":
+		return addressAddReplaceDel(c)
+	case "flush":
+		return addressFlush()
+	default:
+		return fmt.Errorf(" subcommand %s not yet implemented, expected: %v", c, whatIWant)
```
[the remark this answers is not in the document]
Comment (RiSKeD) [reply cmds/core/ip/address_linux.go:32]: Good call, i will check the parsing for each command again later on.

File: cmds/core/ip/parsing_linux.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+		return cmds[x]
+	}
+
+	return ""
+}
+
+var ErrNotFound = fmt.Errorf("not found")
+
+// in the ip command, turns out 'dev' is a noise word.
+// The BNF it shows is not right in that case.
```
[the remark this answers is not in the document]
Comment (RiSKeD) [reply cmds/core/ip/parsing_linux.go:67]: fixed

**Review by rminnich — COMMENTED** [review]
This is really impressive. Nice tests. I'm a bit overwhelmed at the size of it all, but it seems like an improvement ...
And you're not printing a BNF on error :)

**Comment by RiSKeD** [comment]
I know this is a lot. I've refactored the parsing into some convenient helper functions, these are used all over the sub-commands. For now, this PR has exhausted all the current Netlink library's functionalities to my knowledge.

**Review by rminnich — COMMENTED** [review]
at some point, the old and the new commands got broken
./ip.risked link set
2024/07/30 09:22:45 ip: args: [./ip.risked link set], I got to arg 0, expected [] after that

we know it did not expect [] after that.
standard command says this: 
ip link set
Not enough information: "dev" argument is required.
(which is a real improvement: it no longer dumps the bnf either)

but the standard command behavior is terrible:
 ip link set dev
Command line is not complete. Try option "help"
it should show me the possibilities.

this branch shows
./ip.risked link set dev
2024/07/30 09:25:31 ip: args: [./ip.risked link set dev], I got to arg 0, expected [] after that
which is definitely wrong; it should show a slice of strings of network interfaces.

This is a lot closer but still not quite what we want.

**Comment by RiSKeD** [comment]
> at some point, the old and the new commands got broken ./ip.risked link set 2024/07/30 09:22:45 ip: args: [./ip.risked link set], I got to arg 0, expected [] after that
> 
> we know it did not expect [] after that. standard command says this: ip link set Not enough information: "dev" argument is required. (which is a real improvement: it no longer dumps the bnf either)
> 
> but the standard command behavior is terrible: ip link set dev Command line is not complete. Try option "help" it should show me the possibilities.
> 
> this branch shows ./ip.risked link set dev 2024/07/30 09:25:31 ip: args: [./ip.risked link set dev], I got to arg 0, expected [] after that which is definitely wrong; it should show a slice of strings of network interfaces.
> 
> This is a lot closer but still not quite what we want.

Something went wrong when I refactored the globals into the cmd struct and forgot to use Pointer receiver funcs instead of value-based funcs. I'll push a commit with a fix

**Comment by rminnich** [comment]
will test when I get home. thanks.

**Review by rminnich — APPROVED** [review]

**Pull request merged**